client/polling/poll.go

```diff
@@ -0,0 +1,154 @@
+
+	// Run until context dies
+	go func() {
+		defer close(outc)
+		err := pollForever(ctx, logger, c, height, outc)
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+			return
+		}
+		logger.Error("failed polling for events", "err", err)
+	}()
+	return outc, nil
+}
+
+
+
+func pollForever(ctx context.Context, logger log.Logger, c client.Client, startHeight int64, output chan<- abci.ResponseDeliverTx) error {
+	const extraDelay = 100.0 // milliseconds
+	const blockPeriod = time.Second * 5
+	currentHeight := startHeight
+	requiredDelay := blockPeriod
+
+	// compute the sliding average of the time between blocks, starting at double the block period
+	avg := newSlidingAverage(2*blockPeriod.Seconds(), 20)
+
+	var lastBlock *coretypes.ResultBlock
+
+	for {
+		if requiredDelay > 0 {
+			delay := time.After(requiredDelay)
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-delay:
+			}
+		}
+
+		block, err := c.Block(ctx, &currentHeight)
+		if err != nil {
+			// Check to see if the block was requested too soon
+			if strings.Contains(err.Error(), magicForBlockDoesNotExistErr) {
+				// Wait a little bit longer, then ask for the block again
+				requiredDelay = blockPeriod / 4
+				requiredDelay += time.Duration(rand.Intn(250)) * time.Millisecond
+				continue
+			}
+			logger.Error("could not retrieve block", "height", currentHeight, "err", err)
+
+			requiredDelay = blockPeriod * 2
+			requiredDelay += time.Duration(rand.Intn(250)) * time.Millisecond
+			continue
+		}
+		currentHeight++
+
+		txs := block.Block.Txs
+		failureCount := 0
+		for _, txn := range txs {
+			gotTransaction := false
+			for !gotTransaction {
+				txResult, err := c.Tx(ctx, txn.Hash(), false)
+				if err != nil {
+					logger.Error("could not retrieve transaction from block", "height", block.Block.Height, "tx", txn.String())
+
+					retryDelay := baseDelayBetweenTransactions
+					if failureCount != 0 {
+						retryDelay += 3 * time.Millisecond * time.Duration(math.Pow(1.5, float64(failureCount)))
+					}
+					if retryDelay > maxDelayBetweenTransactions {
+						retryDelay = maxDelayBetweenTransactions
+					}
+
+					retryDelay += baseDelayBetweenTransactions * time.Duration(rand.Float64())
+
+					if failureCount < math.MaxInt32 {
+						failureCount += 1
+					}
+
+					select {
+					case <-time.After(retryDelay):
+					case <-ctx.Done():
+						return ctx.Err()
+					}
+					continue
+				}
+				gotTransaction = true
+				select {
+				case output <- txResult.TxResult:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+			}
+		}
+
+		now := time.Now()
+
+		// Update the sliding average
+		if nil != lastBlock {
+			blockTime := block.Block.Time.Sub(lastBlock.Block.Time)
+			avg.push(blockTime.Seconds())
+		}
+		nextBlockAt := block.Block.Time.Add(time.Millisecond * time.Duration(extraDelay+(1000.0*avg.getAverage())))
+		requiredDelay = nextBlockAt.Sub(now) // may be negative
+		lastBlock = block
```

Comment:
honestly I think that querying every second is fine.  this comes from the indexer, not an abci module

Reply:
On the provider, I would guess that querying frequency is probably irrelevant as the RPC node should be dedicated & local. When the CLI is running however, it could actually cause quite a bit more traffic to an RPC node